File: Java/src/_0038_Count_and_Say/Solution.java
package _0038_Count_and_Say;

import java.util.HashMap;
import java.util.Map;

/**
 * Date: 2020/10/11 10:02
 * Content:
 */
public class Solution {

    public String countAndSay(int n) {

        Map<Integer, String> map = new HashMap<Integer, String>() {{
            put(1, "1");
            put(2, "11");
            put(3, "21");
            put(4, "1211");
            put(5, "111221");
        }};
        if (n <= 5) return map.get(n);

        for (int i = 6; i <= n; i++) {
            String s = map.get(i - 1);
            int j = 0;
            StringBuilder sb = new StringBuilder();
            while (j < s.length()) {

                char c = s.charAt(j);
                int k = j + 1;
                while (k < s.length() && c == s.charAt(k)) k++;
                int count = k - j;
                sb.append(count);
                sb.append(c);
                j = k;
            }
            map.put(i, sb.toString());
        }

        return map.get(n);
    }

    public static void main(String[] args) {

        System.out.println((new Solution()).countAndSay(6));
    }
}

File: Python/_0038_Count_and_Say/Solution.py
# coding=utf-8
# Date: 2021/5/22 10:27
from typing import List


# 执行用时：64 ms, 在所有 Python3 提交中击败了39.46%的用户
# 内存消耗：14.9 MB, 在所有 Python3 提交中击败了71.18%的用户
class Solution:
    def countAndSay(self, n: int) -> str:
        s = "1"
        for i in range(n - 1):
            t = ""
            j = 0
            while j < len(s):
                k = j + 1
                while k < len(s) and s[k] == s[j]:
                    k += 1
                t += str(k - j) + str(s[j])
                j = k
            s = t
        return s


if __name__ == "__main__":
    print(Solution().countAndSay(1))  # 1
    print(Solution().countAndSay(2))  # 11
    print(Solution().countAndSay(3))  # 21
